test/FrequencyMapAgentTest.cpp

```diff
@@ -135,17 +149,76 @@ void FrequencyMapAgentTest::SetUp()
     m_agent = geopm::make_unique<FrequencyMapAgent>(*m_platform_io, *m_platform_topo);
     m_num_policy = m_agent->policy_names().size();
 
-    EXPECT_CALL(*m_platform_io, control_domain_type("CPU_FREQUENCY_MAX_CONTROL"));
     EXPECT_CALL(*m_platform_topo, num_domain(GEOPM_DOMAIN_BOARD));
-    EXPECT_CALL(*m_platform_io, push_signal("REGION_HASH", _, _));
-    EXPECT_CALL(*m_platform_io, push_control("CPU_FREQUENCY_MAX_CONTROL", _, _));
+    EXPECT_CALL(*m_platform_io, control_domain_type("CPU_FREQUENCY_MAX_CONTROL"));
+    EXPECT_CALL(*m_platform_io, push_signal("REGION_HASH", _, _)).Times(M_NUM_CPU);
+    EXPECT_CALL(*m_platform_io, push_control("CPU_FREQUENCY_MAX_CONTROL", _, _)).Times(M_NUM_CPU);
     EXPECT_CALL(*m_platform_io, push_control("CPU_UNCORE_FREQUENCY_MIN_CONTROL", _, _));
     EXPECT_CALL(*m_platform_io, push_control("CPU_UNCORE_FREQUENCY_MAX_CONTROL", _, _));
     EXPECT_CALL(*m_platform_io, read_signal("CPU_FREQUENCY_MIN_AVAIL", _, _));
     EXPECT_CALL(*m_platform_io, read_signal("CPU_FREQUENCY_MAX_AVAIL", _, _));
     EXPECT_CALL(*m_platform_io, read_signal("CPU_UNCORE_FREQUENCY_MIN_CONTROL", _, _));
     EXPECT_CALL(*m_platform_io, read_signal("CPU_UNCORE_FREQUENCY_MAX_CONTROL", _, _));
 
+
+}
+
+//Sets up GPU expecations and performs init. Every test must call this.
+void FrequencyMapAgentTest::setup_gpu(int num_gpu)
+{
+    std::set<std::string> control_names = {
+        "CPU_FREQUENCY_MAX_CONTROL",
+        "CPU_UNCORE_FREQUENCY_MIN_CONTROL",
+        "CPU_UNCORE_FREQUENCY_MAX_CONTROL"};
+
+    if (num_gpu > 0) {
+        ON_CALL(*m_platform_io, control_domain_type("GPU_CORE_FREQUENCY_MAX_CONTROL"))
+            .WillByDefault(Return(GEOPM_DOMAIN_GPU));
+        control_names.insert({
+            "GPU_CORE_FREQUENCY_MAX_CONTROL",
+            "GPU_CORE_FREQUENCY_MIN_CONTROL"});
+    }
+    else {
+        ON_CALL(*m_platform_io, control_domain_type("GPU_CORE_FREQUENCY_MAX_CONTROL"))
+            .WillByDefault(Return(GEOPM_DOMAIN_INVALID));
```

Comment:
```
.WillOnce(Throw(std::runtime_error("injected error")));
```
We have code like this in other tests that may be useful to use here.  It should also work for WillRepeatedly, etc.